### ISSUE TITLE
[CCC] System of rules and format of results

### ISSUE THREAD

**Issue by grandsong** [jneen/parsimmon#153]
## Purpose

"CCC" is an abbreviation for "**common confusion and chores**".

If an issue/requirement is agreed to be a CCC issue, it is most preferable to **be solved by the official** (here, you, @wavebeem) so that many users will be saved; they can quickly move forward and focus on their real, valuable problems per project.

Otherwise, users have to "re-invent their own wheels" since there is no agreement for what are common in problems (and hence solutions).

This post is a continuation and restart of #151.

## Summary

Here is a list of interrelating **questions** which I think to be common **confusion**.

- Does the order of rules in my code (of the same file) matter? Why so?
- How to deal with recursive reference?
- Is `lazy` a no-brain solution? And is it in fact the preferable practice?
- What are the most common patterns of rules? (Hopefully covers 80% of common form of languages.)
- And are there good way to shorten them?
- If it is bad to include rule name in some results, then why and when?
- Or is it actually OK to always include rule names in all results?
- If yes, then how to automatically store the name of every rule and then "inject" to all its results?
- Or do I have to endure the inconvenience and always manually duplicate the names in each result at the risk of missing, outdated, etc?
- BTW, should I use rule id (number) instead of rule name (string) for the sake of performance?
- Why `P.index` is used like a standalone "token"?
- Is there another way to get index for all results? (There should be, right? I guess so because I see the index data in error messages.)
- Why can't I get these basic data universally before (and then they can be used in) `P.map`?
- Where is (are) the good place/places for codes to manipulate/interfere the parsing process for all rules to meet my needs mentioned above?

Here is a list of interrelating **requirements** which I think to be common **chores**.

- Manage a central **dictionary** to which all rules are **registered** and from which they are **fetched**, with friendly **exception/error handlers**.
- Support **free order** when writing rules.
- Support **recursive reference** among rules in a good and yet fool-proof way.
- **Shorthands** to define rules quickly and concisely as long as they follow some **common patterns**.
- **All** rules have **names** which are bond to their rules and thus can be **used** when produce parsing results **automatically**.
- **All** parsing results can include **`index`** (position by line and column).
- The texts consumed are recored as **"literal" properties** of results. (Currently, texts are the leaf results themselves)
- **Combine/Compose** the above data into the results of rules when parsing.

The basic demands behind these requirements is signficant ease...

- to write & read codes, 
- to debug,
- to get richer and more useful results, and
- to find and follow good practices (recommended or even forced by official)

## Importance

Headache?

That's what users may feel sooner or later if they go farther than casual use.

Things can be worse if unthoughtful and inexperienced people rush ahead without ever being aware of these questions. They will pay when ugly, poor or bad codes accumulate and unexpected problems ambush them.

It took me quite some time to make a largely fine solution for the requirements and answered most of the questions. Only by having done so can I **settle down and fiddle my grammar efficiently**.

Other users after me will **repeat** similar tasks, which are not really exciting or rewarding.

After a "honeymoon" (happy beginning), they may grow more and more frustrated when the **real start** of the most valuable work on their language grammar is **postponed** again and again.

They would miss **some parser generators who take care of most of above CCC**. However, Parsimmon is too elegant, free and powerful to give up. Thus they will fall in a dilemma.

## A Story of the same trade

Parsimmon is not a parser generator. It shares the [same idea of Chevrotain](https://github.com/SAP/chevrotain/blob/master/docs/faq.md#-why-should-i-use-a-parsing-dsl-instead-of-a-parser-generator).

However, Chevrotain has recently moved to a new phase in which the **default** output will be a format called **CST** (Concrete Syntax Tree).

A little example of a CST node is `{image: "cash", startColumn: 1, startLine: 1, startOffset: 0, endColumn: 4, endLine: 1, endOffset: 3, tokenClassName: "id", tokenType: 53}`. This is the identifier (token name is "id") "cash" in an assignment statement. (Note: property "image" here [just means "literal"](https://github.com/SAP/chevrotain/issues/443).)

Its author @bd82 [explained more about his plan on CST output and CST visitor](https://github.com/SAP/chevrotain/issues/442#issuecomment-292803824). He and I also discussed worries that newbies may have about [impact of CST on performance, which in fact is insignificant](https://github.com/SAP/chevrotain/issues/442#issuecomment-292863549).

Now there is [an example about the use of CST in newest Tutorial Step 3](https://github.com/SAP/chevrotain/blob/master/docs/tutorial/step3_adding_actions_root.md#the-solutions).

I liked Chevrotain. However, its "black magic" inside (which analyzes the raw code text of rule functions and convert them into real yet secret rules) hurt me a lot when I struggled to simplify common rule patterns (which is a much demand due to Chevrotain's lengthy and ugly method names). I gave up at last.

Then I came to Parsimmon. It does much better at what Chevrotain tried but failed to be good at.

I hope it doesn't sound greedy when I request you "borrow" CST, which is a good thing that Chevrotain has now but Parsimmon doesn't.

## Background

The list of parser generators/engines written in JavaScript is short. I have learnt and tried most of them.

Parsimmon is my best choice. And it actually works as easily as I expected!

I am working on a new language which has as many rules as TypeScript. Take a look at [Grammar of TypeScript](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#a-grammar) and you will immediately understand why I desperately need a good way to manage rules systematically and write rules very concisely.

My extremely large (or, insane) project is a rare example. I guess most serious users will want to deal with **middle-scale** languages of similar complexity as [Lua](https://github.com/antlr/grammars-v4/blob/master/lua/Lua.g4). Still, when fiddling rules like "stat" and "exp", which are composed of many sub rule combinations, will **amplify** the current weak (or absent) features of Parsimmon.

## In closing

There is a lot more to say. And I am not sure about answers to a few of the questions (in the common confusion list).

But I'd better end this topic so that it will remain focused.

Please go back and reread the **first section** and the **two lists**.

Then tell me what you think.

Please break your reply into smaller comments, on **a sub group of issues** at a time, so we can make progress step by step **steadily**.

**Comment by wavebeem** [comment]
> Does the order of rules in my code (of the same file) matter? Why so?

Yes the order *can* matter. Variables must be initialized before use in JavaScript, which can cause issues with Parsimmon parsers that take other parsers as arguments. This is why `lazy` exists.

> How to deal with recursive reference?

Use `lazy`.

> Is `lazy` a no-brain solution? And is it in fact the preferable practice?

I personally think `lazy` is an unfortunate thing to have to use.

> What are the most common patterns of rules? (Hopefully covers 80% of common form of languages.)
> And are there good way to shorten them?

I only really know from personal experience of writing a few different parsers, I don't have knowledge of everyone else's code. As for shortening I did write a tips section in the API documentation talking about how to make some common things shorter.

> If it is bad to include rule name in some results, then why and when?
> Or is it actually OK to always include rule names in all results?
> If yes, then how to automatically store the name of every rule and then "inject" to all its results?

This is a hard question to answer. I think generally you want to include a description on a parser which corresponds to a "token". So like "string" or "number" or "identifier", but not to an entire "for" loop. I've been struggling with how to make this more clear.

> Or do I have to endure the inconvenience and always manually duplicate the names in each result at the risk of missing, outdated, etc?
> BTW, should I use rule id (number) instead of rule name (string) for the sake of performance?
> Why `P.index` is used like a standalone "token"?
> Is there another way to get index for all results? (There should be, right? I guess so because I see the index data in error messages.)
> Why can't I get these basic data universally before (and then they can be used in) `P.map`?
> Where is (are) the good place/places for codes to manipulate/interfere the parsing process for all rules to meet my needs mentioned above?

You can do something like this:

```js
var forLoop =
  Parsimmon.seqMap(
    Parsimmon.index,
    Keyword("for"),
    Identifier,
    Keyword("in"),
    Expr,
    Block,
    Parsimmon.index,
    (begin, _, iter, _, expr, block, end) => {
      return someAstNode(...);
    }
  )
```

Honestly `seqMap` is probably one of the most useful functions in Parsimmon. Pretty much every step in your parser can be a `seqMap` with `Parsimmon.index` at the beginning and end to gather the line/column information.

* * *

> Manage a central **dictionary** to which all rules are **registered** and from which they are **fetched**, with friendly **exception/error handlers**.
> Support **free order** when writing rules.
> Support **recursive reference** among rules in a good and yet fool-proof way.

As mentioned in the previous issue, I think my "parsimmon salad" idea could be helpful here for organizing a language and lifting the mental burden of laziness/rule ordering/file splitting.

> **Shorthands** to define rules quickly and concisely as long as they follow some **common patterns**.

I need a more clear request here in order for this to be actionable. I'm not against the spirit of it but I don't really know what's best right now.

> **All** rules have **names** which are bond to their rules and thus can be **used** when produce parsing results **automatically**.
> **All** parsing results can include **`index`** (position by line and column).
> The texts consumed are recored as **"literal" properties** of results. (Currently, texts are the leaf results themselves)
> **Combine/Compose** the above data into the results of rules when parsing.

So if I understand correctly what you want is for Parsimmon to include a way to automatically generate the CST from a parse for you, rather than manually doing it in every case, right? I definitely agree that this results in a lot of busywork, and it's something I have written my own abstractions for already in my personal projects.

My only problem with this request is I have a hard time seeing how to implement this without completely breaking API compatibilitiy for Parsimmon. Maybe it could be done on the "parsimmon salad" level? I would have to play around and see how this particular problem could work within the scope of Parsimmon.

**Comment by grandsong** [comment]
I guess you have agreed that nearly all the questions and problems in my lists are CCC and thus they need to be solved by the official (you).

The purpose of this post is already fulfilled.

But there come a new questions:

- Why do you say 'lazy is an unfortunate thing'? What's the downside of using `lazy`?

And here are some questions & problems that you didn't seem to understand me.

- Shortcut aliases such as `var P = Parsimmon;` is necessary, but has nothing to do with shortening codes with common rule patterns. Moverover, you don't need to know "enough" common patterns to have confidence that you can provide an additional but very help feature which can be extended by you and users in the future. I'll talk about this in a new post.
- The name of a rule is not a mere description. It is identical to the key of the rule in the rules dictionary. And it will be used later in AST/CST/whatever for semantic analysis/validation/etc.
- I know how to use `P.index` in `P.seqMap`. The problem is that I don't like to do so repeatedly for every rule.

---

Basically you seem to have difficulty with making further progress than your little (as what you've shown me) salad.

I will write proposals as solutions in other posts.

**Comment by wavebeem** [comment]
The problem with `lazy` is that it's not always clear when you need to use it, but it's clunky to manually use it on *every* parser.

If you have specific suggestions on how to improve Parsimmon's API I'm happy to hear them.

I get the feeling you want a much higher level API for Parsimmon. If you could write some example of code of what you think would be good, that could be helpful.

**Comment by wavebeem** [comment]
Closing this as discussion is now in https://github.com/jneen/parsimmon/issues/155
